teps/0010-optional-workspaces.md

```diff
@@ -0,0 +1,367 @@
+.ssh directory or a .gitconfig file. These could be modeled as two separate
+optional Workspaces.
+2. Credentials can sometimes be safely omitted. E.g. git does not need
+credentials to clone a public repo like Tekton Pipelines.
+3. Tekton currently has multiple competing mechanisms for mounting files,
+specifically credential files, into a Task: creds-init, Workspaces,
+volumes/volumeMounts. Writing a Catalog Task to support multiple mechanisms
+implies each can be optionally omitted.
+
+### Supplying optional data to inform a Task's behaviour
+
+Allow Tasks to accept an optional set of data that will change the operation
+of that Task in some way.
+
+Examples of this include:
+- accepting an optional linter configuration to change the warnings/errors that are
+raised from the defaults.
+- accepting an optional kubeconfig file that allows a Task to target an external
+cluster instead of the same cluster it's running in.
+- accepting an optional manifest naming specific files to be uploaded instead of
+simply uploading an entire directory of content.
+
+### Receiving optional data generated by a Task
+
+Allow Tasks to _optionally_ generate some data and supply that to the user
+or other Tasks in a Pipeline.
+
+Examples of usage include:
+- Receiving test results as a file as well as printed to stdout.
+- Receiving a manifest of file paths that were copied to external storage.
+- Receiving a diff of files that were modified during Task execution.
+
+All of the above are examples of behaviours that could be opt-in: a user/pipeline
+may not need the results of a test run beyond a pass/fail exit code. A team may
+not need to care which specific files were copied out to external storage; only
+that the transfer succeeded. The diff of modifications made during execution may
+have no bearing on subsequent Tasks in a Pipeline.
+
+A Task can also conceivably skip some portion of work if the user has elected
+not to receive the generated data. For example, a Task may not need to compute
+the diff of its modifications or generate a manifest of changed files if the
+user has not provided a location for that data to be received.
+
+### Goals
+
+1. Allow Tasks to declare optional support for file-based credentials.
+2. Enable Tasks to offer support of Tekton's multiple competing credential mechanisms.
+3. Allow Tasks to optionally receive storage with some data to use during execution.
+4. Allow Tasks to optionally receive storage on to which the Task will place some output data.
+5. Enable Tasks to detect if any declared optional storage has been provided or not at runtime.
+
+### Non-Goals
+
+1. Declaring / binding specific file content on the optional storage.
+2. Validation of the file contents of the optional storage.
+3. Validation of "competing" filesystem fields used in a single TaskRun. An optional Workspace
+and an optional PipelineResource competing for the same location on the filesystem will be left
+up to the runtime to handle.
```

Comment:
i dont totally understand this last one, maybe you could add an example?

Reply:
Does it make more sense with the added knowledge that PipelineResources can already be declared `optional` today?  So a PipelineResource declared `optional` and a Workspace declared `optional` could conceivably "fight" for the same location on disk.

Reply:
(if they're both provided by a user)

Reply:
yes! i didnt realize that, thanks :D

> will be left up to the runtime to handle.

im not 100% clear what "the runtime" means in this case - the controller? kubernetes itself?

Reply:
Ah, good catch I'll update this to be clearer.  Yeah I meant "Kubernetes or whatever system is implementing the Tekton API".  I guess maybe `platform` might be a better identifier here?

Reply:
haha and I've confused myself.  I do quite specifically mean Kubernetes in this case :- k8s will receive a Pod with two volumeMounts both mounting at the same place (one for the resource and one for the workspace).  I'm saying it's just kinda up to kubernetes to figure out how or if that causes problems.

Reply:
are you saying that in this case, the pipelines controller would try to create a pod that mounts the workspace to the specified location, and then the pipelineresource may try to mount to it as well, and itll be up to the underlying platform to decide what to do?

(if so, i wonder in this case if it would be a better experience to notice this in the pipeline controller and treat the run as invalid?)

Reply:
yeah, that's i'm saying this TEP wouldn't protect the user against.  I can add validation for that as a requirement and remove this.

Reply:
ah kk thanks for explaining. i think it makes sense as a requirement so sgtm, but plz push back if you disagree!